crates/ruff_linter/src/rules/pydocstyle/rules/ends_with_punctuation.rs

```diff
@@ -103,13 +103,13 @@ pub(crate) fn ends_with_punctuation(checker: &mut Checker, docstring: &Docstring
         if !trimmed.ends_with(['.', '!', '?']) {
             let mut diagnostic = Diagnostic::new(EndsInPunctuation, docstring.range());
             // Best-effort fix: avoid adding a period after other punctuation marks.
-            if !trimmed.ends_with([':', ';']) {
+            if !trimmed.ends_with([':', ';', '?', '!']) {
                 diagnostic.set_fix(Fix::unsafe_edit(Edit::insertion(
                     ".".to_string(),
                     line.start() + trimmed.text_len(),
                 )));
+                checker.diagnostics.push(diagnostic);
             }
```

Comment:
```suggestion
            }
            checker.diagnostics.push(diagnostic);
```

Reply:
I think we still need to move the `checker.diagnostic.push` call outside the `trimmed.ends_with` branch or we omit diagnostics for docstring summaries that end in `!`. We actually see this because I would expect to see more diagnostics in `ruff_linter__rules__pydocstyle__tests__D415_D415.py.snap`

Reply:
> I think we still need to move the `checker.diagnostic.push` call outside the `trimmed.ends_with` branch or we omit diagnostics for docstring summaries that end in `!`. We actually see this because I would expect to see more diagnostics in `ruff_linter__rules__pydocstyle__tests__D415_D415.py.snap`

Does it look any better now?